README.md
# pofmt

[![Tests](https://github.com/frostming/pofmt/workflows/Tests/badge.svg)](https://github.com/frostming/pofmt/actions?query=workflow%3Aci)
[![pypi version](https://img.shields.io/pypi/v/pofmt.svg)](https://pypi.org/project/pofmt/)
[![pre-commit](https://img.shields.io/badge/pre--commit-enabled-brightgreen?logo=pre-commit&logoColor=white)](https://github.com/pre-commit/pre-commit)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

Your missing PO formatter and linter

## Features

- Wrap msgid and msgstr with a constant max width.
- Can act as a [pre-commit](https://pre-commit.com/) hook.
- Display lint errors in a human readable format.
- Insert white-spaces between Chinese characters and latin letters with [pangu](https://github.com/vinta/pangu.py)
- The widths of CJK characters are multiplied by a factor for visual balance.

## Requirements

pofmt requires Python >=3.6

## Installation

It is recommended to install with `pipx`, if `pipx` haven't been installed yet, refer to the [pipx's docs](https://github.com/pipxproject/pipx)

```bash
$ pipx install pofmt
```

Alternatively, install with `pip` to the user site:

```bash
$ python -m pip install --user pofmt
```

If you are formatting PO files with Chinese, it is recommended to install `pofmt[zh]`. This includes
a handy function to add spaces between CJK characters and latin letters.

## As a pre-commit hook

See [pre-commit](https://pre-commit.com/) for instructions.

A sample `.pre-commit-config.yaml`:

```yaml
- repo: https://github.com/frostming/pofmt
  rev: '0.1.0'
  hooks:
    - id: pofmt
      additional_dependencies: ['pangu']  # for handling Chinese documents
```

## Usage

```
USAGE: pofmt [-h] [--line-length LINE_LENGTH] [-c] [--cjk-width CJK_WIDTH] [filename ...]

Format PO files for consistency

positional arguments:
  filename              Filenames to format, default to all po files under the current directory(recursively)

optional arguments:
  -h, --help            show this help message and exit
  --line-length LINE_LENGTH
                        The max length of msgid and msgstr
  -c, --check           Check only, don't modify files
  --cjk-width CJK_WIDTH
                        The width factor of a CJK character, default: 1.8
  --no-msgid            Don't format msgid
```

## Sample output

```diff
--- Original
+++ Current
@@ -21,49 +21,48 @@
 msgid "Welcome to Flask"
 msgstr "欢迎来到 Flask 的世界"

-msgid "Flask: web development, one "drop" at a time"
+msgid "Flask: web development, one \"drop\" at a time"
 msgstr "Flask： Web 开发，一次一滴"

 #: ../../index.rst:11
 msgid ""
-"Welcome to Flask's documentation. Get started with :doc:`installation` "
-"and then get an overview with the :doc:`quickstart`. There is also a more"
-" detailed :doc:`tutorial/index` that shows how to create a small but "
+"Welcome to Flask's documentation. Get started with :doc:`installation` and"
+" then get an overview with the :doc:`quickstart`. There is also a more "
+"detailed :doc:`tutorial/index` that shows how to create a small but "
 "complete application with Flask. Common patterns are described in the "
 ":doc:`patterns/index` section. The rest of the docs describe each "
 "component of Flask in detail, with a full reference in the :doc:`api` "
 "section."
 msgstr ""
-"欢迎来到Flask的文档。你可以从 :doc:`installation` 入手，然后阅读:doc:`quickstart`来了解基本概念。还有一个包含更多细节的:doc:`tutorial/index`"
-"介绍如何用Flask创建一个很小但是完整的程序。一般的开发模式可以在:doc:`patterns/index`章节找到。剩下的文档详细的介绍了Flask的每一个组成部件，"
-"其中:doc:`api`章节包括完整的API参考信息。"
+"欢迎来到 Flask 的文档。你可以从 :doc:`installation` 入手，然后阅读:doc:`quickstart` "
+"来了解基本概念。还有一个包含更多细节的:doc:`tutorial/index` 介绍如何用 Flask "
+"创建一个很小但是完整的程序。一般的开发模式可以在:doc:`patterns/index` 章节找到。剩下的文档详细的介绍了 Flask "
+"的每一个组成部件，其中:doc:`api` 章节包括完整的 API 参考信息。"

 #: ../../index.rst:19
-msgid "Flask depends on the `Jinja`_ template engine and the `Werkzeug`_ WSGI toolkit. The documentation for these libraries can be found at:"
-msgstr ""
-"Flask 依赖 `Jinja`_ 模板引擎和 `Werkzeug`_ WSGI 工具集。这些库的文档如下："
+msgid ""
+"Flask depends on the `Jinja`_ template engine and the `Werkzeug`_ WSGI "
+"toolkit. The documentation for these libraries can be found at:"
+msgstr "Flask 依赖 `Jinja`_ 模板引擎和 `Werkzeug`_ WSGI 工具集。这些库的文档如下："

 #: ../../index.rst:22
 msgid "`Jinja documentation <https://jinja.palletsprojects.com/>`_"
-msgstr "`Jinja文档<https://jinja.palletsprojects.com/>`_"
+msgstr "`Jinja 文档 <https://jinja.palletsprojects.com/>`_"

 #: ../../index.rst:23
 msgid "`Werkzeug documentation <https://werkzeug.palletsprojects.com/>`_"
-msgstr "`Werkzeug文档<https://werkzeug.palletsprojects.com/>`_"
+msgstr "`Werkzeug 文档 <https://werkzeug.palletsprojects.com/>`_"

 #: ../../index.rst:30
 msgid "User's Guide"
-msgstr ""
-"用户指南"
+msgstr "用户指南"

 #: ../../index.rst:32
 msgid ""
 "This part of the documentation, which is mostly prose, begins with some "
 "background information about Flask, then focuses on step-by-step "
 "instructions for web development with Flask."
-msgstr ""
-"这部分的文档大部分是独立章节，以一些关于 Flask 的背景信息开始，然后重点介绍如何"
-"使用 Flask 一步步进行 Web 开发。"
+msgstr "这部分的文档大部分是独立章节，以一些关于 Flask 的背景信息开始，然后重点介绍如何使用 Flask 一步步进行 Web 开发。"

 #: ../../index.rst:66
 msgid "API Reference"
@@ -73,13 +72,13 @@
 msgid ""
 "If you are looking for information on a specific function, class or "
 "method, this part of the documentation is for you."
-msgstr ""
-"如果你想找关于某个特定函数、类或方法的信息，那么这部分文档就是为你准备的。"
+msgstr "如果你想找关于某个特定函数、类或方法的信息，那么这部分文档就是为你准备的。"

 #: ../../index.rst:78
 msgid "Additional Notes"
 msgstr "附加笔记"

 #: ../../index.rst:80
-msgid "Design notes, legal information and changelog are here for the interested."
+msgid ""
+"Design notes, legal information and changelog are here for the interested."
 msgstr "如果你感兴趣的话，这里有一些设计笔记、法律信息和变更日志（changelog）。"
```


CHANGELOG.md
# Changelog

<!-- insertion marker -->
[v0.3.0](https://github.com/frostming/pofmt/releases/tag/0.3.0) (2021-06-09)
----------------------------------------------------------------------------

### Features & Improvements

- Add an option `--no-msgid` to turn off the reformatting of msgid. [#3](https://github.com/frostming/pofmt/issues/3)


[v0.2.2](https://github.com/frostming/pofmt/releases/tag/0.2.2) (2021-06-09)
----------------------------------------------------------------------------

### Features & Improvements

- Split at CJK punctuations. [#2](https://github.com/frostming/pofmt/issues/2)


[v0.2.1](https://github.com/frostming/pofmt/releases/tag/0.2.1) (2021-06-08)
----------------------------------------------------------------------------

No significant changes.


[v0.2.0](https://github.com/frostming/pofmt/releases/tag/0.2.0) (2021-06-08)
----------------------------------------------------------------------------

### Features & Improvements

- Apply a width factor on CJK characters when measuring the length. Add an option `--cjk-width` to adjust the value. [#1](https://github.com/frostming/pofmt/issues/1)




tests/__init__.py
"""Tests suite for `pofmt`."""

from pathlib import Path

TESTS_DIR = Path(__file__).parent
FIXTURES_DIR = TESTS_DIR / "fixtures"


tests/test_cli.py
import shutil
from pathlib import Path

from pofmt.core import cli

FIXTURES = Path(__file__).parent / "fixtures"


def test_find_po_files(tmp_path, capfd):
    golden = FIXTURES / "golden.po"
    dest_paths = [tmp_path / "a.po", tmp_path / "deep/inner/path/b.po"]
    for path in dest_paths:
        path.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(FIXTURES / "raw.po", path)

    result = cli([str(tmp_path)])
    assert result != 0
    out, _ = capfd.readouterr()
    assert "Checked 2 file(s)" in out
    assert "2 file(s) changed" in out
    for path in dest_paths:
        assert path.read_text("utf-8") == golden.read_text("utf-8")


def test_passing_glob_pattern(tmp_path, capfd):
    golden = FIXTURES / "golden.po"
    raw = FIXTURES / "raw.po"
    dest_paths = [tmp_path / "a.po", tmp_path / "deep/inner/path/b.po"]
    for path in dest_paths:
        path.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(FIXTURES / "raw.po", path)

    result = cli([f"{tmp_path}/*.po"])
    assert result != 0
    out, _ = capfd.readouterr()
    assert "Checked 1 file(s)" in out
    assert "1 file(s) changed" in out

    assert dest_paths[0].read_text("utf-8") == golden.read_text("utf-8")
    assert dest_paths[1].read_text("utf-8") == raw.read_text("utf-8")


def test_check_without_update(tmp_path, capfd):
    raw = FIXTURES / "raw.po"
    dest_paths = [tmp_path / "a.po", tmp_path / "deep/inner/path/b.po"]
    for path in dest_paths:
        path.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(FIXTURES / "raw.po", path)

    result = cli(["--check", str(tmp_path)])
    assert result != 0
    out, _ = capfd.readouterr()
    assert "Need update" in out
    assert "Checked 2 file(s)" in out
    assert "2 file(s) changed" in out

    for path in dest_paths:
        assert path.read_text("utf-8") == raw.read_text("utf-8")


tests/test_formatter.py
import textwrap
from pathlib import Path

import pytest

from pofmt.core import ParseError, Source

FIXTURES = Path(__file__).parent / "fixtures"


@pytest.mark.parametrize("n", range(10))
def test_format_file_consistent(n):
    """Run several times to see if the content is consistent"""
    golden = FIXTURES / "golden.po"
    raw = FIXTURES / "raw.po"
    s = Source(raw)
    assert s.fix(76)
    content = "\n".join(s.lines) + "\n"
    assert content == golden.read_text(encoding="utf-8")


def test_format_file_with_3x_cjk_width():
    golden = FIXTURES / "golden_3x_width.po"
    raw = FIXTURES / "raw.po"
    s = Source(raw)
    assert s.fix(76, 3.0)
    content = "\n".join(s.lines) + "\n"
    assert content == golden.read_text(encoding="utf-8")


def test_format_file_no_msgid():
    golden = FIXTURES / "golden_no_msgid.po"
    raw = FIXTURES / "raw.po"
    s = Source(raw)
    assert s.fix(76, no_msgid=True)
    content = "\n".join(s.lines) + "\n"
    assert content == golden.read_text(encoding="utf-8")


@pytest.mark.parametrize(
    "error_text",
    [
        textwrap.dedent(
            """msgid
    msgstr "hello"
    """
        ),
        textwrap.dedent(
            """msgid "hello
    msgstr "world"
    """
        ),
        'msgstr "hello world"',
        textwrap.dedent(
            """msgid ""
        hello world"
        msgstr "foobar"
        """
        ),
    ],
)
def test_parse_error(error_text):
    s = Source("test_file", error_text.splitlines())
    with pytest.raises(ParseError):
        s.parse()


src/pofmt/core.py
import argparse
import builtins
import codecs
import contextlib
import difflib
import glob
import locale
import os
import re
import sys
import textwrap
import typing as t
import unicodedata
from pathlib import Path

try:
    import pangu
except ModuleNotFoundError:
    pangu = None

MESSAGE_RE = r'"(.*)"[ ]*$'
_cjk_opening_punct = (
    r"[\uff08\u3008\u300a\u300c\u300e\ufe43\u3014\uffe5\u3010\u201c\u2018]"
)
_cjk_closing_punct = (
    r"[\uff09\u3009\u300b\u300d\u300f\ufe44\u3015\u2026\u2014\uff5e\ufe4f"
    r"\u3001\u3002\uff0c\uff1f\uff01\uff1a\uff1b\u201d\u2019]"
)


class ParseError(Exception):
    pass


def escape_quotes(text: str) -> str:
    return re.sub(r'(?<!\\)"', '\\"', text)


def is_full_width(text: str) -> bool:
    """See https://stackoverflow.com/a/31666966"""
    return unicodedata.east_asian_width(text) in "WAF"


def support_unicode():
    """Check whether operating system supports main symbols or not."""
    encoding = getattr(sys.stdout, "encoding")
    if encoding is None:
        encoding = locale.getpreferredencoding(False)

    try:
        encoding = codecs.lookup(encoding).name
    except Exception:
        encoding = "utf-8"
    return encoding == "utf-8"


@contextlib.contextmanager
def len_with_cjk_width_factor(factor: float) -> t.Generator[None, None, None]:
    builtin_len = builtins.len

    def new_len(text: str) -> int:
        if not isinstance(text, str):
            return builtin_len(text)
        return int(sum(factor if is_full_width(c) else 1 for c in text))

    builtins.len = new_len
    try:
        yield
    finally:
        builtins.len = builtin_len


class Span(t.NamedTuple):
    start: int
    end: int


class Entry:
    def __init__(self, span: Span, msgid: t.List[str], msgstr: t.List[str]) -> None:
        self.span = span
        self.msgid = msgid
        self.msgstr = msgstr

    @staticmethod
    def process_text(lines: t.List[str]) -> str:
        text = escape_quotes("".join(lines))
        if pangu is not None:
            text = pangu.spacing_text(text)
        return text

    def _format_text(
        self, title: str, lines: t.List[str], width: int, reformat: bool = True
    ) -> t.List[str]:
        if not reformat:
            return [f'{title} "{lines[0]}"'] + [f'"{line}"' for line in lines[1:]]

        text = self.process_text(lines)
        if len(title) + len(text) + 3 <= width:
            # 1 space + 2 quotes = 3
            return [f'{title} "{text}"']

        wrapper = textwrap.TextWrapper(width - 2, drop_whitespace=False)
        wrapper.wordsep_re = re.compile(
            wrapper.wordsep_re.pattern.rstrip(") \n")
            + r")|(?<=%s) (?=\w)|(?=%s)))" % (_cjk_closing_punct, _cjk_opening_punct),
            re.VERBOSE,
        )
        return [f'{title} ""'] + [f'"{line}"' for line in wrapper.wrap(text)]

    def format(
        self, width: int, cjk_width_factor: float = 1.8, no_msgid: bool = False
    ) -> t.List[str]:
        with len_with_cjk_width_factor(cjk_width_factor):
            return self._format_text(
                "msgid", self.msgid, width, reformat=not no_msgid
            ) + self._format_text("msgstr", self.msgstr, width)


class Source:
    def __init__(
        self, filename: str, lines: t.Optional[t.Sequence[str]] = None
    ) -> None:
        self.filename = filename
        if lines is None:
            lines = [line for line in Path(filename).read_text("utf-8").splitlines()]
        self.lines = lines
        self._original = self.lines[:]
        self.lineno = -1
        self._entries: t.List[Entry] = []

    def __iter__(self) -> t.Iterator[str]:
        return self

    def __next__(self) -> str:
        self.lineno += 1
        if self.lineno >= len(self.lines):
            raise StopIteration()
        return self.lines[self.lineno]

    def parse_error(self, message: str, lineno: t.Optional[int] = None) -> str:
        if lineno is None:
            lineno = self.lineno
        raise ParseError(f"line {lineno}: {message}")

    def parse(self) -> None:
        if self.lineno >= 0:
            raise RuntimeError("Can't parse multiple times on one source")
        for line in self:
            if line.startswith("#, ") and line[3:].strip() == "fuzzy":
                # Don't modify the fuzzy entries
                self._parse_entry()
            if not line.strip() or line.startswith("#"):
                continue
            elif line.startswith("msgid"):
                self.lineno -= 1
                self._entries.append(self._parse_entry())
            else:
                self.parse_error("Unexpected token")

    def _parse_entry(self) -> Entry:
        msgid, msgstr = [], []
        temp = []
        start_line = self.lineno

        for line in self:
            if not line.strip() or line.startswith("#"):
                break
            if line.startswith("msgid"):
                if msgid:
                    self.lineno -= 1
                    break
                match = re.match(MESSAGE_RE, line[6:])
                if not match:
                    self.parse_error('Expect `msgid "..."`')
                msgid.append(match.group(1))
                start_line = self.lineno
                temp = msgid
            elif line.startswith("msgstr"):
                if msgstr:
                    self.lineno -= 1
                    break
                match = re.match(MESSAGE_RE, line[7:])
                if not match:
                    self.parse_error('Expect `msgstr: "..."`')
                msgstr.append(match.group(1))
                temp = msgstr
            else:
                match = re.match(MESSAGE_RE, line)
                if not match:
                    self.parse_error('Expect `"..."`')
                temp.append(match.group(1))

        if not msgstr:
            self.parse_error("Missing msgstr")
        return Entry(Span(start_line, self.lineno), msgid, msgstr)

    def fix(
        self,
        line_length: int,
        cjk_width: float = 1.8,
        no_msgid: bool = False,
        show: bool = False,
    ) -> bool:
        self.parse()
        for entry in reversed(self._entries):
            self.lines[entry.span.start : entry.span.end] = entry.format(
                line_length, cjk_width, no_msgid
            )
        return self.diff(show)

    def diff(self, show: bool = False) -> bool:
        has_diff = False
        show_title = False
        for line in difflib.unified_diff(
            self._original, self.lines, "Original", "Current", lineterm=""
        ):
            if show:
                if not show_title:
                    print(f"Need update: {self.filename}")
                    show_title = True
                print(line)
            has_diff = True
        return has_diff

    def write(self, path: t.Union[str, Path]) -> None:
        Path(path).write_text("\n".join(self.lines) + "\n", encoding="utf-8")


def cli(argv: t.Optional[t.Sequence[str]] = None) -> int:
    parser = argparse.ArgumentParser(description="Format PO files for consistency")
    parser.add_argument(
        "--line-length", type=int, default=76, help="The max length of msgid and msgstr"
    )
    parser.add_argument(
        "-c", "--check", action="store_true", help="Check only, don't modify files"
    )
    parser.add_argument(
        "--cjk-width",
        type=float,
        default=1.8,
        help="The width factor of a CJK character, default: 1.8",
    )
    parser.add_argument("--no-msgid", action="store_true", help="Don't format msgid")
    parser.add_argument(
        "filename",
        nargs="*",
        default=[os.getcwd()],
        help="Filenames to format, default to all po files under "
        "the current directory(recursively)",
    )
    args = parser.parse_args(argv)

    identical, changed, errors = 0, 0, 0
    if support_unicode():
        ERROR, SUCCESS = "❌", "✨"
    else:
        ERROR, SUCCESS = ":(", ":)"

    for filename in args.filename:
        if os.path.isdir(filename):
            filename = os.path.join(filename, "**/*.po")
        for path in glob.glob(filename, recursive=True):
            source = Source(path)
            try:
                if source.fix(
                    args.line_length,
                    cjk_width=args.cjk_width,
                    no_msgid=args.no_msgid,
                    show=args.check,
                ):
                    if not args.check:
                        source.write(path)
                        print(f"{SUCCESS} {path} is updated")
                    changed += 1
                else:
                    identical += 1
            except ParseError as e:
                errors += 1
                print(f"{ERROR} {path} Parse error: {e}")
                continue

    print(
        f"\nChecked {identical + changed + errors} file(s), "
        f"{errors} error file(s) and {changed} file(s) changed."
    )
    if changed or errors:
        return 1
    return 0
